### DISCUSSION TITLE
Closures in API routes cause memory leaks?

### DISCUSSION THREAD

**Discussion by fibonacid** [vercel/next.js#21863] [Help]
Hello, i was wondering if is it safe to cache data using closures on API routes.
Consider the following example:

```javascript
let cache;

async function fetchSomeData() {
  const response = await fetch("http://some-api.com/data.json");
  return response.json();
}

async function handle(req, res) {
  cache = cache ?? (await fetchSomeData());

  res.json(cache);
}

export default handle;

```

I'm doing something similar on a project but i'm getting this error:
`MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 SIGINT listeners added.`

What is the best approach to create API routes that depend on async data ?
Something like top level await would be nice, but I don't know if it's supported.

**Comment by leerob** [comment]
Use `cache-control` headers!

Here’s an example API Route that sends a JSON response and caches that response for one day: 

```js
// pages/api/user.js

export default function handler(req, res){
  res.setHeader(
    'Cache-Control',
    's-maxage=86400'
  );
  res.status(200).json({name: ‘John Doe’ });
};
```

As for the memory leak, it sounds like you're not reusing your connection to whatever service you're connecting to. Try using a singleton.

**Reply by fibonacid** [reply]
Thanks for the reply, your blog is like the bible :)

I'm not sure the `Cache-Control` approach works in my case.
Basically i'm using an API route as a GraphQL stitching gateway, so i need to cache the remote executable schema only once.

I tried the singleton approach by creating a GatewayClient class.

```javascript
// pages/api/graphql.js

export default async handler(req, res)  {
  const gateway = new GatewayClient();
  const schema = await gateway.getSchema();

  const apolloServer = new ApolloServer({ schema });
  const middleware = apolloServer.createHandler({ path: "/api/graphql" });

  return middleware(req, res);
};
```

This seems to work, although it looks a bit complex.
Is it possible to wait for some data before the api function is launched by Next ?

**Reply by leerob** [reply]
That's still creating a new `GatewayClient` every time, since it's inside the serverless function body. Have you tried the Apollo examples in the `/examples` folder?

**Reply by fibonacid** [reply]
> Have you tried the Apollo examples in the /examples folder?

Yeah i had a look, that's where i have found the `createHandler` method.

> That's still creating a new GatewayClient every time

GatewayClient should be implemented as a singleton as you said

Would you personally handle this task using an API route or extending the Next.js server ?
